go/src/staleFlagCleaner.go

```diff
@@ -0,0 +1,624 @@
+	// store it in a hashmap
+	for _, element := range sfc.methods.Methods {
+		sfc.functype[element.MethodName] = element
+	}
+}
+
+// Gets the type of the flagAPI
+func (sfc *staleFlagCleaner) flagTypeAPI(callExpr *dst.CallExpr) API {
+	var apiName string
+	var flagIndex int // -1 if flag or group not here
+	flagIndex = -1
+
+	for ind, expr := range callExpr.Args {
+		switch arg := expr.(type) {
+		case *dst.Ident:
+			if arg.Name == sfc.flagName {
+				flagIndex = ind
+				break
+			}
+		}
+	}
+
+	switch d := callExpr.Fun.(type) {
+	case *dst.Ident:
+		apiName = d.Name
+	case *dst.SelectorExpr:
+		apiName = d.Sel.Name
+	default:
+		// other cases will come in future
+	}
+
+	if element, ok := sfc.functype[apiName]; ok {
+		if element.FlagIndex == flagIndex {
+			switch element.OfType {
+			case treated:
+				return isTreated
+			case control:
+				return isControl
+			default:
+				return isUnknown
+			}
+		}
+	}
+	return isUnknown
+}
+
+/* This function will evaluate all expression nodes here */
+func (sfc *staleFlagCleaner) evaluateExprNode(exprNode dst.Expr) Value {
+
+	switch exprUnderConsideration := exprNode.(type) {
+	case *dst.Ident:
+		if val, ok := sfc.valueMap[exprUnderConsideration.Name]; ok {
+			return val
+		}
+		switch exprUnderConsideration.Name {
+		case strTrue:
+			return isTrue
+		case strFalse:
+			return isFalse
+		}
+
+	// Handling expressions related to calling a function
+	case *dst.CallExpr:
+		// Get the type of API
+		typeOfAPI := sfc.flagTypeAPI(exprUnderConsideration)
+		// Hardcoding the output of some API types
+		if typeOfAPI == isTreated {
+			if sfc.isTreated {
+				return isTrue
+			}
+			return isFalse
+		}
+		if typeOfAPI == isControl {
+			if sfc.isTreated {
+				return isFalse
+			}
+			return isTrue
+		}
+	//Binary expressions
+	case *dst.BinaryExpr:
+		valX := sfc.evaluateExprNode(exprUnderConsideration.X)
+		valY := sfc.evaluateExprNode(exprUnderConsideration.Y)
+		if valX != isBot || valY != isBot {
+			switch exprUnderConsideration.Op {
+			case token.LAND: //stands for &&
+				if valX == isFalse || valY == isFalse {
+					return isFalse
+				}
+				if valX == isTrue && valY == isTrue {
+					return isTrue
+				}
+			case token.LOR: //stands for ||
+				if valX == isTrue || valY == isTrue {
+					return isTrue
+				}
+				if valX == isFalse && valY == isFalse {
+					return isFalse
+				}
+			case token.EQL:
+				if valX != isBot && valY != isBot {
+					if valX == valY {
+						return isTrue
+					}
+					return isFalse
+				}
+
+			case token.NEQ:
+				if valX == valY {
+					return isFalse
+				}
+				return isTrue
+			}
+		}
+	//Unary expressions
+	case *dst.UnaryExpr:
+		switch exprUnderConsideration.Op {
+		case token.NOT:
+			valX := sfc.evaluateExprNode(exprUnderConsideration.X)
+			switch valX {
+			case isTrue:
+				return isFalse
+			case isFalse:
+				return isTrue
+			}
+		case token.AND: // &expr
+			return sfc.evaluateExprNode(exprUnderConsideration.X)
```

Comment:
Uh? What is this case? If this is actually for `&expr`, it doesn't seem correct to me, since taking the address of a boolean value (e.g. `&true` not a `*bool`) just gives you an error.

Reply:
I am not sure how this works either. Since we simply return `X` here, don't we equate `expr` with `&expr`? What is the reasoning behind it (similar question for the `StarExpr`)?

Reply:
Take an example (I know, it's very rare).....
```
staleVariable := funcAffectedByStaleFlag(.....)  
......
pointerToVar := &staleVariable
a := *pointerToVar && false
```
To handle the below two expressions I did it. I tried extracting the value of the variable from 2nd expression and store it in the `valueMap` which can be used to handle the 3rd expression.